File: library/src/main/java/me/shkschneider/skeleton/helper/AssetsHelper.java
package me.shkschneider.skeleton.helper;

import android.content.res.AssetManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import me.shkschneider.skeleton.data.InternalDataHelper;

public class AssetsHelper {

    protected AssetsHelper() {
        // Empty
    }

    public static AssetManager assetManager() {
        return ContextHelper.applicationContext().getAssets();
    }

    @Nullable
    public static List<String> list() {
        final List<String> list = new ArrayList<>();
        try {
            Collections.addAll(list, assetManager().list(""));
            return list;
        }
        catch (final IOException e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    @Nullable
    public static InputStream open(@NonNull final String name) {
        final AssetManager assetManager = assetManager();
        try {
            return assetManager.open(name);
        }
        catch (final IOException e) {
            LogHelper.wtf(e);
            return null;
        }
    }

    public static boolean dump() {
        int errors = 0;
        final List<String> assets = list();
        if (assets == null) {
            errors++;
        }
        else {
            for (final String asset : assets) {
                try {
                    final InputStream inputStream = open(asset);
                    if (inputStream == null) {
                        throw new NullPointerException();
                    }
                    final OutputStream outputStream = InternalDataHelper.openOutput(asset);
                    if (outputStream == null) {
                        throw new NullPointerException();
                    }
                    byte[] buffer = new byte[1024];
                    int read;
                    while ((read = inputStream.read(buffer)) != -1) {
                        outputStream.write(buffer, 0, read);
                    }
                    inputStream.close();
                    outputStream.flush();
                    outputStream.close();
                }
                catch (final IOException e) {
                    LogHelper.wtf(e);
                    errors++;
                }
            }
        }
        return (errors == 0);
    }

}

File: library/src/main/kotlin/me/shkschneider/skeleton/helper/AssetsHelper.kt
package me.shkschneider.skeleton.helper

import android.content.res.AssetManager
import me.shkschneider.skeleton.data.InternalDataHelper
import java.io.IOException
import java.io.InputStream

object AssetsHelper {

    fun assetManager(): AssetManager {
        return ContextHelper.applicationContext().assets
    }

    fun list(): List<String>? {
        val assetManager = assetManager()
        try {
            return assetManager.list("")?.toList()
        } catch (e: IOException) {
            Logger.wtf(e)
            return null
        }
    }

    fun open(name: String): InputStream? {
        val assetManager = assetManager()
        try {
            return assetManager.open(name)
        } catch (e: IOException) {
            Logger.wtf(e)
            return null
        }
    }

    fun dump(): Boolean {
        var errors = 0
        val assets = list()
        assets ?: errors++
        assets?.forEach {
            try {
                val inputStream = open(it) ?: throw NullPointerException()
                val outputStream = InternalDataHelper.openOutput(it) ?: throw NullPointerException()
                inputStream.copyTo(outputStream)
                inputStream.close()
                outputStream.flush()
                outputStream.close()
            } catch (e: IOException) {
                Logger.wtf(e)
                errors++
            }
        }
        return errors == 0
    }

}
